### PR TITLE
=htc #20920 create ContentLength HTTP header in JavaDSL

### PR DESCRIPTION
**Pull request by naferx**

### CODE REVIEW

**Comment by naferx** [comment]
Refs #20920

File: akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+package akka.http.javadsl.model.headers;
+
+/**
+ *  Model for the `Content-Length` header.
+ *  Specification: https://tools.ietf.org/html/draft-ietf-httpbis-p1-messaging-26#section-3.3.2
+ */
+public abstract class ContentLength extends akka.http.scaladsl.model.HttpHeader {
+
+    public static ContentLength create(Long length) {
```
Comment (ktoso) [inline akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java]: primitive `long` please :)

**Comment by ktoso** [comment]
Excellent, thanks a lot!

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
LGTM

**Comment by ktoso** [comment]
Thanks! Will merge after validation :)

**Comment by naferx** [comment]
Thank you too!

**Comment by 2beaucoup** [comment]
Shouldn't the constructor stay `private[akka]`?

**Comment by ktoso** [comment]
> Shouldn't the constructor stay private[http]?
 
WDYM? it still is.

**Comment by 2beaucoup** [comment]
It's now exposed via the Java API.

**Comment by ktoso** [comment]
Note that the JavaDSL class is written in Java, so we can't `private[http]` there.
Maybe making it protected would signal intent somewhat (though doesnt forbid people to abuse and new it anyway...) WDYT?

**Comment by 2beaucoup** [comment]
But why should the static public constructor be added at all?

**Comment by ktoso** [comment]
Hah, you confused me by using the word `constructor` there for a sec 😉 
You mean the `create` factory method... you're right, we don't want it.

File: akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+package akka.http.javadsl.model.headers;
+
+/**
+ *  Model for the `Content-Length` header.
+ *  Specification: https://tools.ietf.org/html/draft-ietf-httpbis-p1-messaging-26#section-3.3.2
+ */
+public abstract class ContentLength extends akka.http.scaladsl.model.HttpHeader {
+
+    public static ContentLength create(long length) {
```
Comment (ktoso) [inline akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java]: as @2beaucoup mentioned, we don't want people to actually create these – it's a special header that is created/calculated by Akka internally. Please remove the create method :)

https://github.com/akka/akka/pull/20926#issuecomment-231596467

**Comment by 2beaucoup** [comment]
Should have commented inline. Thx for bearing with me.

File: akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java (hunk 2 above)
Comment (naferx) [inline akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java]: PR updated

**Comment by ktoso** [comment]
PLS BUILD

File: akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.http.javadsl.model.headers;
+
+/**
+ *  Model for the `Content-Length` header.
+ *  Specification: https://tools.ietf.org/html/draft-ietf-httpbis-p1-messaging-26#section-3.3.2
+ */
+public abstract class ContentLength extends akka.http.scaladsl.model.HttpHeader {
```
Comment (ktoso) [inline akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java]: this is javadsl, thus it should extend `akka.http.scaladsl.model.HttpHeader`

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala [hunk 4]
```diff
@@ -362,7 +362,8 @@ object `Content-Length` extends ModeledCompanion[`Content-Length`]
  * Instances of this class will only be created transiently during header parsing and will never appear
  * in HttpMessage.header. To access the Content-Length, see subclasses of HttpEntity.
  */
-final case class `Content-Length` private[http] (length: Long) extends RequestResponseHeader {
+final case class `Content-Length` private[http] (length: Long) extends jm.headers.ContentLength
+  with RequestResponseHeader {
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: good

File: akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java (hunk 3 above)
Comment (2beaucoup) [inline akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java]: All headers in `akka.http.javadsl.model.headers` seem to extend `akka.http.scaladsl.model.HttpHeader`.

Comment (ktoso) [inline akka-http-core/src/main/java/akka/http/javadsl/model/headers/ContentLength.java]: Heh, my bad for reviewing without referring to current impl.

**Comment by ktoso** [comment]
Seems we're good to go here, thanks a lot @naferx !

**Pull request merged**